### PR TITLE
feat: implement blobToSha256

### PR DESCRIPTION
**Pull request by sk-tenba**
https://app.asana.com/0/home/1203849283056222/1204115146622395

### CODE REVIEW

**Comment by sk-tenba** [comment]
Screenshot from 2023-03-22 12-04-26

File: modules/with-bufs-as-converters/codecs/CodecBlob.ts [hunk 1]
```diff
@@ -1,23 +1,32 @@
...
+  if (blob.size < 64 << 20) {
+    const buf = await blob.arrayBuffer();
+    const sha256 = await arrayBufferToSha256(buf);
+    return { data: "sha256:" + sha256, bufs: { [sha256]: blob } };
+  } else {
+    const sha256 = await blobToSha256(blob);
+    return { data: "sha256:" + sha256, bufs: { [sha256]: blob } };
+  }
+}
+
```
Comment (sk-kitsune) [inline modules/with-bufs-as-converters/codecs/CodecBlob.ts:24]: The if-else logic should not be put in this file. It has to be put under `blobToSha256`.

---

🔴 **Blocking issue**: this PR cannot be merged until this issue is resolved.

Comment (sk-kitsune) [reply modules/with-bufs-as-converters/codecs/CodecBlob.ts:24]: extract `64 << 20` to a constant

Comment (sk-kitsune) [reply modules/with-bufs-as-converters/codecs/CodecBlob.ts:24]: where is the lazy-load?

**Review by sk-kitsune — CHANGES_REQUESTED** [review]
⛔ **Changes requested**: there are blocking issues.

File: modules/with-bufs-as-converters/utils/buffer.ts [hunk 2]
```diff
@@ -25,3 +27,23 @@ export async function arrayBufferToSha256(
   const bytesOfSha256 = await crypto.subtle.digest("SHA-256", buffer);
   return arrayBufferToHex(bytesOfSha256);
 }
+
+export async function blobToSha256(blob: Blob): Promise<Sha256> {
+  /**NOTE: @sk-tenba: for better performance, if the size is less than 64MB, the hash
+   * the SHA-256 should be calculated from the array buffer.
+   */
+  const crypto = await import("crypto");
```
Comment (sk-kitsune) [inline modules/with-bufs-as-converters/utils/buffer.ts:35]: This module is imported in all cases.

It should only be loaded if `blob.size >= SUBTLE_CRYPTO_SIZE_LIMIT`

---

🔴 **Blocking issue**: this PR cannot be merged until this issue is resolved.

Comment (sk-kitsune) [reply modules/with-bufs-as-converters/utils/buffer.ts:35]: simply place this statement after the if statement

File: modules/with-bufs-as-converters/utils/buffer.ts [hunk 3]
```diff
@@ -25,3 +27,23 @@ export async function arrayBufferToSha256(
...
+  if (blob.size < SUBTLE_CRYPTO_SIZE_LIMIT) {
```
Comment (sk-kitsune) [inline modules/with-bufs-as-converters/utils/buffer.ts:36]:
```
if (blob.size < SUBTLE_CRYPTO_SIZE_LIMIT && globalThis?.crypto?.subtle != null)
```

Comment (sk-kitsune) [reply modules/with-bufs-as-converters/utils/buffer.ts:36]: 🔴 **Blocking issue**: this PR cannot be merged until this issue is resolved.

Comment (sk-kitsune) [reply modules/with-bufs-as-converters/utils/buffer.ts:36]: btw, we can remove function `arrayBufferToSha256` by copying its code to `blobToSha256`

**Review by sk-kitsune — APPROVED** [review]
✅ **Approved**: no blocking issues

**Pull request merged**